src/signing/collector/signatures_collector.rs

```diff
@@ -0,0 +1,554 @@
+                break; // finished early, we have fulfilled signing requirements of all transactions
+            }
+            self.sign_with_factors_of_kind(factor_sources_of_kind)
+                .await?;
+        }
+        Ok(())
+    }
+
+    fn input_for_interactor(
+        &self,
+        factor_source_id: &FactorSourceIDFromHash,
+    ) -> BatchTXBatchKeySigningRequest {
+        self.state
+            .borrow()
+            .petitions
+            .borrow()
+            .input_for_interactor(factor_source_id)
+    }
+
+    fn request_for_serial_interactor(
+        &self,
+        factor_source_id: &FactorSourceIDFromHash,
+    ) -> SerialBatchSigningRequest {
+        let batch_signing_request = self.input_for_interactor(factor_source_id);
+
+        SerialBatchSigningRequest::new(
+            batch_signing_request,
+            self.invalid_transactions_if_skipped(factor_source_id)
+                .into_iter()
+                .collect_vec(),
+        )
+    }
+
+    fn request_for_parallel_interactor(
+        &self,
+        factor_source_ids: IndexSet<FactorSourceIDFromHash>,
+    ) -> ParallelBatchSigningRequest {
+        let per_factor_source = factor_source_ids
+            .clone()
+            .iter()
+            .map(|fid| (*fid, self.input_for_interactor(fid)))
+            .collect::<IndexMap<FactorSourceIDFromHash, BatchTXBatchKeySigningRequest>>();
+
+        let invalid_transactions_if_skipped =
+            self.invalid_transactions_if_skipped_factor_sources(factor_source_ids);
+
+        // Prepare the request for the interactor
+        ParallelBatchSigningRequest::new(per_factor_source, invalid_transactions_if_skipped)
+    }
+
+    fn invalid_transactions_if_skipped(
+        &self,
+        factor_source_id: &FactorSourceIDFromHash,
+    ) -> IndexSet<InvalidTransactionIfSkipped> {
+        self.state
+            .borrow()
+            .petitions
+            .borrow()
+            .invalid_transactions_if_skipped(factor_source_id)
+    }
+
+    fn invalid_transactions_if_skipped_factor_sources(
+        &self,
+        factor_source_ids: IndexSet<FactorSourceIDFromHash>,
+    ) -> IndexSet<InvalidTransactionIfSkipped> {
+        factor_source_ids
+            .into_iter()
+            .flat_map(|f| self.invalid_transactions_if_skipped(&f))
+            .collect::<IndexSet<_>>()
+    }
+
+    fn process_batch_response(
+        &self,
+        response: SignWithFactorSourceOrSourcesOutcome<BatchSigningResponse>,
+    ) {
+        let state = self.state.borrow_mut();
+        let petitions = state.petitions.borrow_mut();
+        petitions.process_batch_response(response)
+    }
+
+    fn outcome(self) -> SignaturesOutcome {
+        let state = self.state.borrow_mut();
+        let petitions = state.petitions.borrow_mut();
+        let expected_number_of_transactions = petitions.txid_to_petition.borrow().len();
+        drop(petitions);
+        drop(state);
```

Comment:
why explicit drop? `self` will be gone by the end of this function scope

Reply:
because I'm borrowing `state` twice, If I remove the drop we get:
<img width="586" alt="Screenshot 2024-08-28 at 10 05 59" src="https://github.com/user-attachments/assets/f62b6f22-f278-4646-b4ae-726025272b0a">

and I've tried with a scope, did not get that to work. I think this is the correct usage of `drop`..?

Reply:
because I'm borrowing `state` twice, If I remove the drop we get:

<img width="586" alt="Screenshot 2024-08-28 at 10 05 59" src="https://github.com/user-attachments/assets/f62b6f22-f278-4646-b4ae-726025272b0a">

and I've tried with a scope, did not get that to work. I think this is the correct usage of `drop`..?

Reply:
You mean that something like this does not work?
```rust
let expected_number_of_transactions;
{
    let state = self.state.borrow_mut();
    let petitions = state.petitions.borrow_mut();
    expected_number_of_transactions = petitions.txid_to_petition.borrow().len();
}
```

Reply:
Ah that works! I tried *this* which does **not** work:

```rust
let expected_number_of_transactions = {
    let state = self.state.borrow_mut();
    let petitions = state.petitions.borrow_mut();
    petitions.txid_to_petition.borrow().len()
};
```

TIL...

Reply:
https://github.com/radixdlt/sargon-mfa/commit/a1d0cdcf94ef84ea8e82b5ecc2156d50332d757d :)